### PR TITLE
Memcached Receiver Skeleton

### PR DESCRIPTION
**Pull request by keitwb**
This is the start of a basic memcached receiver.  It scrapes some basic
metrics from a memcached instance and sends them along.  The primary
test of the metric generation logic is in the integration test due to
the relative opacity of the memcached protocol and the lack of any
preexisting mock (the integration test uses a real memcached instance
via Docker containers).

There are only five basic metrics to start with but this can be easily
expanded with this framework.

### CODE REVIEW

**Comment by jrcamp** [comment]
Can you use the new scraper helper? https://github.com/open-telemetry/opentelemetry-collector/blob/015827f6ed982d15c043319c7c367f9431664178/receiver/receiverhelper/scraper.go (It's not been used widely yet so would be good to give it a test drive and see if it needs any tweaking)

File: receiver/memcachedreceiver/config.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	"go.opentelemetry.io/collector/config/configmodels"
+)
+
+type config struct {
+	configmodels.ReceiverSettings `mapstructure:",squash"`
+
+	// The target memcached host.
+	Endpoint string `mapstructure:"endpoint"`
+
+	Timeout time.Duration `mapstructure:"timeout"`
```
Comment (anuraaga) [inline receiver/memcachedreceiver/config.go:29]:
```suggestion
        // Timeout for requests to fetch memcache stats.
	Timeout time.Duration `mapstructure:"timeout"`
```

File: receiver/memcachedreceiver/factory.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+func createDefaultConfig() configmodels.Receiver {
+	return &config{
+		ReceiverSettings: configmodels.ReceiverSettings{
+			TypeVal: typeStr,
+			NameVal: typeStr,
+		},
+		CollectionInterval: 10 * time.Second,
+		Timeout:            10 * time.Second,
```
Comment (anuraaga) [inline receiver/memcachedreceiver/factory.go:48]: Should the default timeout be at least a little less than the interval to prevent concurrent requests?

**Review by anuraaga — APPROVED** [review]
Just some small points, thanks!

File: receiver/memcachedreceiver/config.go (hunk 1 above)
Comment (anuraaga) [reply receiver/memcachedreceiver/config.go:29]: Sorry you probably shouldn't actually apply the suggestion due to spaces vs tabs but just copy :)

File: receiver/memcachedreceiver/factory.go (hunk 2 above)
Comment (keitwb) [reply receiver/memcachedreceiver/factory.go:48]: Not necessarily because the for/select loop + time.Ticker smooths things out and prevents concurrent requests.

**Comment by keitwb** [comment]
> Can you use the new scraper helper?

What advantage does it provide, especially when there is only one collection loop?  If the ticker loop is desirable to reuse, we can do something like `utils.RunOnInterval` that was in the Smart Agent.

**Comment by tigrannajaryan** [comment]
@keitwb @bogdandrutu is this review still in progress?

**Comment by keitwb** [comment]
Yes, switched it to use the scraper interface.

I still think you are going to be  far better on long term by just offering helpers that you can compose into the receivers (e.g. the config options, the timer sharing thing, and whatever else) instead of forcing inheritance through the scraper interface (that thing is basically an abstract class with the ScraperBase interface being the abstract methods that must be implemented).  But I'm on to a different team soon, so I'll just share my opinion on it and let you all decide how to proceed with Scrapers and the porting of the monitors.

File: cmd/otelcontribcol/components.go [hunk 3]
```diff
@@ -103,6 +104,7 @@ func components() (component.Factories, error) {
 		splunkhecreceiver.NewFactory(),
 		statsdreceiver.NewFactory(),
 		wavefrontreceiver.NewFactory(),
+		memcachedreceiver.NewFactory(),
```
[the remark this answers is not in the document]
Comment (keitwb) [reply cmd/otelcontribcol/components.go]: Removed from components.

**Comment by tigrannajaryan** [comment]
Tests failed.

**Comment by keitwb** [comment]
Coverage does not take into account the integration test.

File: internal/common/testing/container/container.go [hunk 4]
```diff
@@ -45,20 +45,28 @@ type Containers struct {
...
-	ports        nat.PortMap
+	cont         *types.ContainerJSON
 	t            *testing.T
 }
 
-// AddrForPort returns the endpoint with host:port for the container port.
-func (c *Container) AddrForPort(port uint16) string {
-	for dockerPort, bindings := range c.ports {
+// AddrForPort returns the host and port for the container port.
+func (c *Container) AddrForPort(port uint16) (string, uint) {
```
Comment (bogdandrutu) [inline internal/common/testing/container/container.go:53]: Please split this change in a separate PR (everything in internal/common)

File: receiver/memcachedreceiver/config.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/receiver/receiverhelper"
+)
+
+type config struct {
+	configmodels.ReceiverSettings            `mapstructure:",squash"`
+	receiverhelper.ScraperControllerSettings `mapstructure:",squash"`
+
+	// The target memcached host.
+	Endpoint string `mapstructure:"endpoint"`
```
Comment (bogdandrutu) [inline receiver/memcachedreceiver/config.go:29]: Consider to use one of the addresses from `confignet` if possible?

File: receiver/memcachedreceiver/config.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+	// The timeout for the memcache stats request
```
Comment (bogdandrutu) [inline receiver/memcachedreceiver/config.go:31]: s/The timeout/Timeout

File: receiver/memcachedreceiver/config.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/receiver/receiverhelper"
+)
+
+type config struct {
+	configmodels.ReceiverSettings            `mapstructure:",squash"`
+	receiverhelper.ScraperControllerSettings `mapstructure:",squash"`
+
+	// The target memcached host.
```
Comment (bogdandrutu) [inline receiver/memcachedreceiver/config.go:28]: [Go comments](https://golang.org/doc/effective_go.html#commentary) should start with the name of the exported var/func/struct/etc.

File: internal/common/testing/container/container.go (hunk 4 above)
Comment (keitwb) [reply internal/common/testing/container/container.go]: Removed it as it isn't very essential.

File: receiver/memcachedreceiver/config.go (hunk 5 above)
Comment (keitwb) [reply receiver/memcachedreceiver/config.go]: Done.

File: receiver/memcachedreceiver/config.go (hunk 6 above)
Comment (keitwb) [reply receiver/memcachedreceiver/config.go]: Done.

File: receiver/memcachedreceiver/config.go (hunk 7 above)
Comment (keitwb) [reply receiver/memcachedreceiver/config.go]: Done.

**Review by bogdandrutu — APPROVED** [review]
